modules/cart/src/Form/AddToCartForm.php

```diff
@@ -136,9 +132,21 @@ public function getFormId() {
     if ($this->operation != 'default') {
       $form_id = $form_id . '_' . $this->operation;
     }
-    $form_id .= '_' . self::$formInstanceId;
+    $id = sha1(serialize($this->entity->getPurchasedEntity()->toArray()));
```

Comment:
Thank you for the point because I didn't know about this function before. Though for the first look it might seem more concise and 2.15 times faster than sha1(serialize($this->entity->getPurchasedEntity()->toArray())) but, .... it does not work. Despite all the generated IDs may appear unique they are not actually. It's the same issue as with the counter which we have now. The reason why these both methods to generate unique ID for the form don't work is that they operate on the values which exist only in the current PHP memory and when happen AJAX replacement (after choosing another attribute field option) the previous values are not more relevant to objects where they were generated before. PHP description of the spl_object_hash(): 

> This id can be used as a hash key for storing objects, or for identifying an object, **as long as the object is not destroyed**.

Additionally to the above spl_object_hash() ignores the properties of the object which also prevents to uniquely identify it. I think that there is no any other way to do it except relying on the values (properties) saved in the consistent memory.

Note, all the tests are done not only with multiple Add to cart forms but also with duplicated carts on a page.